File: ios/BFExplorer/Plaoc/IPCCore/native/NativeIpc.swift
//
//  NativeIpc.swift
//  BFExplorer
//
//  Created by ui03 on 2023/2/25.
//

import UIKit
import AsyncHTTPClient

class NativeIpc: Ipc {

    private var port: NativePort<IpcMessage, IpcMessage>?
    
    init(port: NativePort<IpcMessage, IpcMessage>, remote: MicroModuleInfo, role: IPC_ROLE) {
        super.init()
        self.port = port
        self.remote = remote
        self.role = role.rawValue
        
        self.supportRaw = true
        self.supportBinary = true
        
        _ = port.onMessage { message in
            
            self.messageSignal?.emit((message,self))
        }
        Task {
            port.start()
        }
    }
    
    override func toString() -> String {
        return super.toString() + "@NativeIpc"
    }
    
    override func doPostMessage(data: IpcMessage) {
        port?.postMessage(msg: data)
    }
    
    override func doClose() {
        port?.close()
    }
}

File: android/MicroService/src/main/java/org/dweb_browser/microservice/ipc/NativeIpc.kt
package org.dweb_browser.microservice.ipc

import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus
import org.dweb_browser.helper.ioAsyncExceptionHandler
import org.dweb_browser.helper.printDebug
import org.dweb_browser.microservice.help.MicroModuleManifest
import org.dweb_browser.microservice.ipc.helper.IPC_ROLE
import org.dweb_browser.microservice.ipc.helper.IpcMessage
import org.dweb_browser.microservice.ipc.helper.IpcMessageArgs

fun debugNativeIpc(tag: String, msg: Any = "", err: Throwable? = null) =
  printDebug("native-ipc", tag, msg, err)

class NativeIpc(
  val port: NativePort<IpcMessage, IpcMessage>,
  override val remote: MicroModuleManifest,
  private val roleType: IPC_ROLE,
) : Ipc() {
  override val role get() = roleType.role
  override fun toString(): String {
    return super.toString() + "@NativeIpc"
  }

  override val supportRaw = true
  override val supportBinary = true

  private val ioAsyncScope = MainScope() + ioAsyncExceptionHandler

  init {
    port.onMessage { message ->
      _messageSignal.emit(IpcMessageArgs(message, this@NativeIpc))
    }
    ioAsyncScope.launch {
      port.onClose { close() }
      port.start()
    }
  }


  override suspend fun _doPostMessage(data: IpcMessage) {
    port.postMessage(data)
  }

  override suspend fun _doClose() {
    port.close()
    ioAsyncScope.cancel()
  }
}
